dali/kernels/imgproc/convolution/laplacian_cpu_test.cc

```diff
@@ -0,0 +1,436 @@
+
+#include <gtest/gtest.h>
+#include <array>
+
+#include "dali/kernels/common/utils.h"
+#include "dali/kernels/imgproc/convolution/laplacian_cpu.h"
+#include "dali/kernels/scratch.h"
+#include "dali/test/tensor_test_utils.h"
+#include "dali/test/test_tensors.h"
+
+namespace dali {
+namespace kernels {
+
+template <int window_size>
+std::array<float, window_size> GetSobelWindow(int d_order) {
+    static_assert(window_size >= 3 && window_size % 2 == 1);
+    std::array<float, window_size> w = {0.};
+    w[0] = 1.;
+    for (int i = 1; i < window_size - d_order; i++) {
+    auto prevval = w[0];
+    for (int j = 1; j < i; j++) {
+      auto val = w[j];
+      w[j] = prevval + w[j];
+      prevval = val;
+    }
+    w[i] = prevval;
+  }
+  for (int i = window_size - d_order; i < window_size; i++) {
+    auto prevval = w[0];
+    w[0] = -prevval;
+    for (int j = 1; j < i; j++) {
+      auto val = w[j];
+      w[j] = prevval - w[j];
+      prevval = val;
+    }
+    w[i] = prevval;
+  }
+  return w;
+}
+
+template <int axes, int window_size>
+void FillKernel(const TensorView<StorageCPU, float, axes>& kernel,
+                 const std::array<std::array<std::array<float, window_size>, axes>, axes>&
+                 windows, int padding) {
+  for (int d = 0; d < axes; d++) {
+    for (int i = 0; i < volume(kernel.shape); i++) {
+      int offset = i;
+      float r = 1.f;
+      for (int j = axes - 1; j >= 0; j--) {
+        int ind = offset % (window_size + padding);
+        ind -= padding / 2;
+        r *= (ind < 0 || ind >= window_size) ? 0.f : windows[d][j][ind];
+        offset /= (window_size + padding);
+      }
+      kernel.data[i] += r;
+    }
+  }
+}
+
+template <int axes_, int window_size_, bool use_smoothing_>
+struct LaplacianWindows {
+  static constexpr int axes = axes_;
+  static constexpr int window_size = window_size_;
+  static constexpr bool use_smoothing = use_smoothing_;
+
+  LaplacianWindows() {
+    for (int i = 0; i < axes; i++) {
+      for (int j = 0; j < axes; j++) {
+        if (i == j) {
+          window_sizes[i][j] = window_size;
+          windows[i][j] = GetSobelWindow<window_size>(2);
+          tensor_windows[i][j] = {windows[i][j].data(), window_size};
+        } else if (use_smoothing) {
+          window_sizes[i][j] = window_size;
+          windows[i][j] = GetSobelWindow<window_size>(0);
+          tensor_windows[i][j] = {windows[i][j].data(), window_size};
+        } else {
+          window_sizes[i][j] = 1;
+          windows[i][j] = uniform_array<window_size>(0.f);
+          auto middle = window_size / 2;
+          windows[i][j][middle] = 1.f;
+          tensor_windows[i][j] = {windows[i][j].data() + middle, 1};
+        }
+      }
+    }
+  }
+  std::array<std::array<int, axes>, axes> window_sizes;
+  std::array<std::array<std::array<float, window_size>, axes>, axes> windows;
+  std::array<std::array<TensorView<StorageCPU, const float, 1>, axes>, axes> tensor_windows;
```

Comment:
Again, I am lost about which level of nesting corresponds to what.

Reply:
It goes the same way as in LaplacianCPU.Run. To recap:
`tensor_windows[i]` describes windows used to compute the `i-th` partial derivative (i.e. the one that approximates the second order partial derivative along `i-th` dimension when counting them from the left to the right). So `tensor_windows[i][i]` is a window that should look like alike [1, -2, 1], whereas for j <> i, `tensor_windows[i][j]` is some kind of a smoothing window.